tlatools/org.lamport.tlatools/src/tla2sany/xml/XMLExporter.java

```diff
@@ -187,12 +196,25 @@ static void moduleToXML(String... args) throws XMLExportingException {
       ExternalModuleTable table = spec.getExternalModuleTable();
       //Element e = table.getRootModule().exportDefinition(doc,context);
       //rootElement.appendChild(e);
-      ModuleNode[] externalModules = table.getModuleNodes();
-      for (int j = 0; j < externalModules.length; j++) {
-        //Element ext_e = externalModules[j].exportDefinition(doc, context);
-        Element ext_e = externalModules[j].export(doc, context);
-        rootElement.appendChild(ext_e);
-      }
+      
+		if (restricted) {
+			final BiPredicate<SemanticNode, SemanticNode> filter = ((s1, s2) -> {
+				if (s1 instanceof OpDefNode && s2 instanceof ModuleNode) {
+					final OpDefNode odn = (OpDefNode) s1;
+					return s2.equals(odn.getOriginallyDefinedInModuleNode());
```

Comment:
Possibly this can be done without the second module parameter using the `OpDefNode.hasSource()` API call, which will return `true` if the operator was defined in another module imported with `INSTANCE`. I have not tested whether it will return `true` if imported using `EXTENDS`.

Reply:
An even better solution would be to refactor the XML conversion code by tearing it all out of these classes into its own class in the `tla2sany.xml` namespace using the Visitor pattern. Then the filter could easily be part of the xml conversion class.

Reply:
The check has been expanded to include `OpDefOrDeclNodes` (VARIABLES and CONSTANTS), which lacks the `hasSource` attribute.